silence/silence.go

```diff
@@ -544,104 +540,179 @@ func validateSilence(s *pb.Silence) error {
 	if s.EndsAt.Before(s.StartsAt) {
 		return errors.New("end time must not be before start time")
 	}
-	if s.UpdatedAt.IsZero() {
-		return errors.New("invalid zero update timestamp")
-	}
 	return nil
 }
 
-// cloneSilence returns a shallow copy of a silence.
-func cloneSilence(sil *pb.Silence) *pb.Silence {
-	s := *sil
-	return &s
+// Expire the silence with the given ID immediately.
+func (s *Silences) Expire(id string) error {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	return s.expire(id)
 }
 
-func (s *Silences) getSilence(id string) (*pb.Silence, bool) {
-	msil, ok := s.st[id]
-	if !ok {
-		return nil, false
+// Set inserts a new silence or updates an existing silence. It inserts a
+// silence when the ID is missing or a silence with the ID does not exist.
+// If updating an existing silence, and the update changes how alerts are
+// matched or the time window in which the silence is active (excluding
+// time extensions), the existing silence is expired and replaced with a
+// new silence with a different ID.
+//
+// When updating an existing silence, the existing silence must be cloned
+// using Clone() before making any modifications to the silence. This
+// prevents situations where the stored silence is modified because both
+// pointers reference the same memory. If Clone() is not used before
+// replacing a silence then the operation will fail.
+//
+// Set returns an error if a silence is invalid, or would exceed any of the
+// existing limits. When this happens, a silence might have an ID even though
+// the operation failed. It is safe to re-use the silence for subsequent calls
+// to Set where previous calls have failed.
+func (s *Silences) Set(sil *pb.Silence) (err error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	now := s.nowUTC()
+	if sil.StartsAt.IsZero() {
+		sil.StartsAt = now
 	}
-	return msil.Silence, true
-}
 
-func (s *Silences) setSilence(sil *pb.Silence, now time.Time, skipValidate bool) error {
-	sil.UpdatedAt = now
+	if err := validateSilence(sil); err != nil {
+		return fmt.Errorf("invalid silence: %w", err)
+	}
 
-	if !skipValidate {
-		if err := validateSilence(sil); err != nil {
-			return fmt.Errorf("silence invalid: %w", err)
+	existing, ok := s.getSilence(sil.Id)
+	// updateExisting is true iff this is an existing silence, and the update does
+	// not affect how alerts are matched or the time window in which the
+	// silence is active (excluding time extensions).
+	updateExisting := ok && canUpdate(existing, sil, now)
+	if !updateExisting {
+		// If this is a new silence, or it is replacing an existing silence,
```

Comment:
This is how I fixed the issue where subsequent calls for the same silence fail even when the validation or limits are fixed. Instead of returning `ErrNotFound`, we give the silence a new ID and turn it into a create operation instead of an update operation.